buildSrc/src/test/java/org/elasticsearch/gradle/precommit/ForbiddenPatternsTaskTests.java

```diff
@@ -0,0 +1,127 @@
+    }
+
+    public void testCheckInvalidPatternsWhenSourceFilesExistNoViolation() throws Exception {
+        Project project = createProject();
+        ForbiddenPatternsTask task = createTask(project);
+
+        String line = "public void bar() {}";
+        File file = new File(project.getProjectDir(), "src/main/java/Foo.java");
+        file.getParentFile().mkdirs();
+        file.createNewFile();
+        Files.write(file.toPath(), line.getBytes());
+
+        task.checkInvalidPatterns();
+
+        File outputMarker = new File(project.getBuildDir(), "markers/forbiddenPatterns");
+        Optional<String> result = Files.readAllLines(outputMarker.toPath(), StandardCharsets.UTF_8).stream().findFirst();
+
+        assertTrue(result.isPresent());
+        assertEquals("done", result.get());
+    }
+
+    public void testCheckInvalidPatternsWhenSourceFilesExistHavingTab() throws Exception {
+        Project project = createProject();
+        ForbiddenPatternsTask task = createTask(project);
+
+        String line = "\tpublic void bar() {}";
+        File file = new File(project.getProjectDir(), "src/main/java/Bar.java");
+        file.getParentFile().mkdirs();
+        file.createNewFile();
+        Files.write(file.toPath(), line.getBytes());
+
+        try {
+            task.checkInvalidPatterns();
+            fail("GradleException was expected to be thrown in this case!");
+        } catch (GradleException e) {
+            assertTrue(e.getMessage().startsWith("Found invalid patterns"));
+        }
+    }
+
+    public void testCheckInvalidPatternsWithCustomRule() throws Exception {
+        Map<String, String> rule = new HashMap<>();
+        rule.put("name", "TODO comments are not allowed");
+        rule.put("pattern", "\\/\\/.*(?i)TODO");
+
+        Project project = createProject();
+        ForbiddenPatternsTask task = createTask(project);
+        task.rule(rule);
+
+        List<String> lines = Arrays.asList("GOOD LINE", "//todo", "// some stuff, toDo");
+        File file = new File(project.getProjectDir(), "src/main/java/Moot.java");
+        file.getParentFile().mkdirs();
+        file.createNewFile();
+        Files.write(file.toPath(), lines, StandardCharsets.UTF_8);
+
+        try {
+            task.checkInvalidPatterns();
+            fail("GradleException was expected to be thrown in this case!");
+        } catch (GradleException e) {
+            assertTrue(e.getMessage().startsWith("Found invalid patterns"));
+        }
+    }
+
+    public void testCheckInvalidPatternsWhenExcludingFiles() throws Exception {
+        Project project = createProject();
+        ForbiddenPatternsTask task = createTask(project);
+        task.exclude("**/*.java");
+
+        File file = new File(project.getProjectDir(), "src/main/java/FooBarMoot.java");
+        file.getParentFile().mkdirs();
+        file.createNewFile();
+        Files.write(file.toPath(), "\t".getBytes());
+
+        task.checkInvalidPatterns();
```

Comment:
might be wroth to extract two methods for the assertions something like `checkAndAssertSuccessfull` and `cehckAndAssertException` - would make the tests more readable.